editor/utils/test/allow-list.js

```diff
@@ -0,0 +1,70 @@
+/**
+ * Internal dependencies
+ */
+import { parseAllowList } from '../allow-list';
+
+describe( 'parseAllowList', () => {
+	it( 'should parse undefined', () => {
+		const allowList = parseAllowList( undefined );
+		expect( allowList ).toEqual( {
+			'*': true,
+		} );
+	} );
+
+	it( 'should parse true', () => {
+		const allowList = parseAllowList( true );
+		expect( allowList ).toEqual( {
+			'*': true,
+		} );
+	} );
+
+	it( 'should parse false', () => {
+		const allowList = parseAllowList( false );
+		expect( allowList ).toEqual( {
+			'*': false,
+		} );
+	} );
+
+	it( 'should parse a string', () => {
+		const allowList = parseAllowList( 'core/verse' );
+		expect( allowList ).toEqual( {
+			'core/verse': true,
+			'*': false,
+		} );
+	} );
+
+	it( 'should parse a negated string', () => {
+		const allowList = parseAllowList( '!core/verse' );
+		expect( allowList ).toEqual( {
+			'core/verse': false,
+			'*': false,
+		} );
+	} );
+
+	it( 'should parse a shorthand array', () => {
+		const allowList = parseAllowList( [ 'core/image', '!core/verse', '*' ] );
```

Comment:
It seems like we're doing a lot of work here to avoid using a JS built-in: a plain object with the block name as a key and `true`/`false` as the allow value. Is this `'!core/verse'` type of argument common elsewhere? It seems to be extra code to run and test to avoid writing `{ 'core/image': false }` and instead write `[ '!core/image' ]`. It doesn't seem worth it to me.

The only benefit is maybe a list of allowed `parseAllowList( [ 'core/image', 'core/verse' ] ) => { 'core/image': true, 'core/verse': true }` but even then I say we just make objects required for where this is used and save a lot of code and make the API more consistent. Consistent APIs are good and lead to fewer bugs 😃

Reply:
Yeah, that makes sense. I'm cool with keeping just `[ 'core/image', 'core/verse' ]`; it's a nice shorthand for allowing a list of blocks.

Reply:
In terms of an easily understandable API I'd just make it another property. There's less room for error that way (it would be easier to spot what's allowed and what's not)... and actually I don't see why they couldn't be combined behind the scenes. Why not:


```js
<WhateverBlock
  allowedBlocks={ [ 'core/image', 'core/verse' ] }
/>
<WhateverBlock
  disallowedBlocks={ [ 'core/video' ] }
/>
/* This should throw an error, because when do you need an allow list and a disallow list when you don't set wildcards? */
<WhateverBlock
  allowedBlocks={ [ 'core/image', 'core/verse' ] }
  disallowedBlocks={ [ 'core/video' ] }
/>
```

That makes it way clearer and you don't have to manually set wildcards.

Reply:
That's how the components would see it... programmatically maybe `parseAllowList` is the wrong approach and it should be `allowList( [ 'core/image', 'core/verse' ] )` and `disallowedList( [ 'core/video' ] )`.

There's no logical need to allow both allowed and disallowed blocks to be explicitly defined in the list... right? 😄

Reply:
> There's no logical need to allow both allowed and disallowed blocks to be explicitly defined in the list... right? 😄

No, and that's part of why I like the object syntax. When typing `{ 'core/video': true }` you're forced to choose `true` or `false` 🙂

Reply:
The only issue with that is there is wildcard for either true or false after everything else, but the object API allows for explicit allowed and disallowed values. But with an allowed list approach, you don’t need to explicitly disallow any values with a `false` setting.

The object api allows folks to set things as both true and false which is confusing. It might be better to accept a flat array as either allowed Or disallowed, and then wildcard everything else as the opposite

Reply:
In a previous version of this PR I tried having only whitelists. The problem with, this, though was that you could get into situations where a parent and child conflict with one another. For example:

```js
// In foo/parent:
<InnerBlocks allowedBlocks={ [ 'foo/child-1' ] }>

// In foo/child-1:
allowedParents: []

// In foo/child-2:
allowedParents: [ 'foo/parent' ]
```

Here, it's not clear what should be allowed.

- Can `foo/child-1` be inserted into `foo/parent`? The parent says yes, but the child says no.
- Can `foo/child-2` be inserted into `foo/parent`? The parent says no, but the child says yes.

This is what motivated me to [introduce the object syntax](https://github.com/WordPress/gutenberg/issues/5540#issuecomment-386192518), which attempts to introduce a distinction between what a block _explicitly_ desires and what it _implicitly_ desires.

```js
// In foo/parent:
<InnerBlocks allowedBlocks={ { 'foo/child-1': true, '*': false } }>

// In foo/child-1:
allowedParents: { '*': false }

// In foo/child-2:
allowedParents: { 'foo/parent': true, '*': false }
```

Now, things are clearer.

- Can `foo/child-1` be inserted into `foo/parent`? The parent explicitly says yes, and the child implicily says no. Explicit wins, so the answer is yes.
- Can `foo/child-2` be inserted into `foo/parent`? The parent implicitly says no, and the child explicitly says yes. Explicit wins, so the answer is yes.

Being able to explicitly say no is useful, too. For example, `foo/parent` might want to prevent `foo/buggy` from ever being nested within it because it causes a nasty styling bug.

```js
// In foo/parent:
<InnerBlocks allowedBlocks={ { 'foo/child-1': true, 'foo/buggy': false, '*': false } }>
```

Here, it doesn't matter what `foo/buggy` specifies—it won't be allowed in `foo/parent`.

😅 Sorry about the long comment. I hope that explains some of my thinking behind the object syntax—let me know what you think.

Reply:
That makes sense... but it's very complex and I'd say tough to explain. You need to know whether something was explicitly/implicitly set to know if it should be allowed.

I find this approach unintuitive because I'd imagine different people will intuit different things from the settings.

Why not just err on the side of "any time something is not allowed: answer is no?" It's much simpler and means there isn't a complex formula to whether a block is allowed. If the parent or the child says no: answer is no.

The example:

```js
<InnerBlocks allowedBlocks={ { 'foo/child-1': true, 'foo/buggy': false, '*': false } }>
```

becomes:

```js
<InnerBlocks allowedBlocks={ [ 'foo/child-1' ] }>
```

As a new developer coming to this API I'd be confused why there are explicit and wildcard `false` values.

Reply:
> Why not just err on the side of "any time something is not allowed: answer is no?" 

Because if we err on the side of 'no', it means that a plugin developer cannot write a block that is insertable into a core block that uses `allowedBlocks` (e.g. [cover image](https://github.com/WordPress/gutenberg/pull/5452/files?utf8=✓&diff=unified#diff-30a5f912f6d7dc9db5d7c49abfced736R187)). This came up in https://github.com/WordPress/gutenberg/issues/5540#issuecomment-380816384.

I do concede that the object syntax is difficult to explain. I like it because of the flexibility it gives plugin developers, but I'd be happy to start with array whitelists (erring on the side of 'yes') and iterate based on developer feedback.